presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceFunction.java

```diff
@@ -0,0 +1,120 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Primitives;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.Signature.typeVariable;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class ArrayReduceFunction
+        extends SqlScalarFunction
+{
+    public static final ArrayReduceFunction ARRAY_REDUCE_FUNCTION = new ArrayReduceFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayReduceFunction.class, "reduce", Type.class, Type.class, Type.class, Block.class, Object.class, MethodHandle.class, MethodHandle.class);
+
+    private ArrayReduceFunction()
+    {
+        super(new Signature(
+                "reduce",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("T"), typeVariable("S"), typeVariable("R")),
+                ImmutableList.of(),
+                parseTypeSignature("R"),
+                ImmutableList.of(parseTypeSignature("array(T)"), parseTypeSignature("S"), parseTypeSignature("function(T,S,S)"), parseTypeSignature("function(S,R)")),
+                false));
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return true;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Reduce elements of the array into a single value";
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type inputType = boundVariables.getTypeVariable("T");
+        Type intermediateType = boundVariables.getTypeVariable("S");
+        Type outputType = boundVariables.getTypeVariable("R");
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(inputType).bindTo(intermediateType).bindTo(outputType);
+        return new ScalarFunctionImplementation(
+                true,
+                ImmutableList.of(false, false, false, false),
+                methodHandle.asType(
+                        methodHandle.type()
+                                .changeParameterType(1, intermediateType.getJavaType())
+                                .changeReturnType(Primitives.wrap(outputType.getJavaType()))),
+                isDeterministic());
+    }
+
+    public static Object reduce(
+            Type inputType,
+            Type intermediateType,
+            Type outputType,
+            Block block,
+            Object initialIntermediateValue,
+            MethodHandle inputFunction,
+            MethodHandle outputFunction)
+    {
+        int positionCount = block.getPositionCount();
+        if (positionCount == 0) {
+            return null;
```

Comment:
@dain and I had a discussion on this after I talked about this behavior.

The choice was made hear to mimic SQL aggregation function behavior. I made a mistake in the original code. I shouldn't treat zero-length array specially. I should instead treat null-only input array (including empty array) specially. I fixed it.

For users who need full control on null element in the input array, the decision we made at the time was that we could introduce a `array_reduce_nullable`.

The choice was also made because the difference in easiness to go from one way to another

* If the defined behavior matches user expection:
  * `array_reduce(arr, initVal, (x, s) -> f(x, s))`
* Otherwise
  * If array_reduce produces null for empty, to produce non-null for null-only input (including empty):
    * `coalesce(array_reduce(arr, initVal, (x, s) -> f(x, s)), initVal)`
  * If array_reduce produces initial value for empty, to produce null for null-only input (including empty):
    * Assuming `f` never produces null:
      * `array_reduce(arr, null, (x, s) -> IF(s IS NULL, f(x, initVal), f(x, s)))`
    * Otherwise:

```
       array_reduce(
         arr,
         CAST(ROW(initVal, false) AS ROW(T val, BOOLEAN hasNonNull)),
         (x, s) -> IF(
           x IS NULL,
           s,
           CAST(ROW(f(x, s), true) AS ROW(T val, BOOLEAN hasNonNull))),
         s -> IF(s.hasNonNull, s.val, null))
```